test/fixtures/list_test_names/decode.bats

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bats
+
+bats_require_minimum_version 1.5.0
+
+bats_encode_test_name() {
+  local name="$1"
+  local result='test_'
+  local hex_code
+
+  if [[ ! "$name" =~ [^[:alnum:]\ _-] ]]; then
+    name="${name//_/-5f}"
+    name="${name//-/-2d}"
+    name="${name// /_}"
+    result+="$name"
+  else
+    local length="${#name}"
+    local char i
+
+    for ((i = 0; i < length; i++)); do
+      char="${name:$i:1}"
+      if [[ "$char" == ' ' ]]; then
+        result+='_'
+      elif [[ "$char" =~ [[:alnum:]] ]]; then
+        result+="$char"
+      else
+        printf -v 'hex_code' -- '-%02x' \'"$char"
+        result+="$hex_code"
+      fi
+    done
+  fi
+
+  printf -v "$2" '%s' "$result"
+}
+
+bats_decode_test_name() {
+  local name="$1"
+  local length="${#name}"
+  name="${name#test_}"
+  local i=0
+  local result=''
+  local char=''
+  while (( i < length )); do
+    char="${name:$i:1}"
+    case $char in
+      '-')
+        (( i += 1 ))
+        if [[ "${name:$i:4}" == '2d5f' ]]; then
+          result+="_"
+          (( i += 4 ))
+        else
+          result+=$(printf "%b" "\x${name:$i:2}")
+          (( i += 2 ))
+        fi
+      ;;
+      '_')
+        result+=" "
+        (( i += 1 ))
+        ;;
+      *)
+        result+="$char"
+        (( i += 1 ))
+    ;;
+    esac
+  done
+  printf -v "$2" '%s' "$result"
+}
```

Comment:
This seems like an awful amount of code duplication to me. Is this really necessary? Can't we just hardcode the expected test results below?

Reply:
First duplication of both functions is due to the fact, that the files where they are defined at bats runtime is not sourceable. This in fact should be eliminated by externalizing the functions to a core library.

If you think about duplication between encode and decode, I would not consider that as a duplicate encode and decode are inverse functions.

The tests written are _black box_ tests not caring about the actual encoded result but just checking that after encode and decode the original name  is returned by decode.

I did not add a test for the _encode_ function itself becuse the function was already there - might be nice to have some there, too
Imho the encode could be tested by defining a hard coded name -> encoded_name list. For the decode I would consider the name ->  encode =>  decode -> name the better approach because it ensures the result is the same.

If we use hardcoded names, a breaking change to the encode function one woul dhave to update *both* not only the tests for encode, but also for decode (at least the list). If not the decode tests would still pass.
using the encode -> decode method will make the decode method fail on braking changes.